google-cloud-core/src/main/java/com/google/cloud/Policy.java

```diff
@@ -157,66 +174,147 @@ protected Builder(Policy policy) {
+          ImmutableList.Builder<String> membersBuilder = ImmutableList.builder();
+          membersBuilder.addAll(binding.getMembers());
+          membersBuilder.add(first.strValue());
+          for (Identity identity : others) {
+            membersBuilder.add(identity.strValue());
+          }
+          bindingBuilder.setMembers(membersBuilder.build());
+          bindingsList.set(i, bindingBuilder.build());
+          return this;
+        }
       }
-      Set<Identity> toAdd = new LinkedHashSet<>();
-      toAdd.add(first);
-      toAdd.addAll(Arrays.asList(others));
-      Set<Identity> identities = bindings.get(checkNotNull(role, "The role cannot be null."));
-      if (identities == null) {
-        identities = new HashSet<>();
-        bindings.put(role, identities);
+      // Binding does not yet exist.
+      Binding.Builder bindingBuilder = Binding.newBuilder().setRole(role.getValue());
+      ImmutableList.Builder<String> membersBuilder = ImmutableList.builder();
+      membersBuilder.add(first.strValue());
+      for (Identity identity : others) {
+        membersBuilder.add(identity.strValue());
       }
-      identities.addAll(toAdd);
+      bindingBuilder.setMembers(membersBuilder.build());
+      bindingsList.add(bindingBuilder.build());
       return this;
     }
 
     /**
      * Removes one or more identities from an existing binding. Does nothing if the binding
      * associated with the provided role doesn't exist.
+     *
+     * @throws IllegalArgumentException if policy version is equal to 3 or has conditional bindings.
      */
     public final Builder removeIdentity(Role role, Identity first, Identity... others) {
-      Set<Identity> identities = bindings.get(role);
-      if (identities != null) {
-        identities.remove(first);
-        identities.removeAll(Arrays.asList(others));
+      checkArgument(
+          !isConditional(this.version, this.bindingsList),
+          "removeIdentity() is only supported with version 1 policies and non-conditional policies");
+      String nullIdentityMessage = "Null identities are not permitted.";
+      checkNotNull(first, nullIdentityMessage);
+      checkNotNull(others, nullIdentityMessage);
+      checkNotNull(role, "The role cannot be null.");
+      for (int i = 0; i < bindingsList.size(); i++) {
+        Binding binding = bindingsList.get(i);
+        if (binding.getRole().equals(role.getValue())) {
+          Binding.Builder bindingBuilder = binding.toBuilder().removeMembers(first.strValue());
+          for (Identity identity : others) {
+            bindingBuilder.removeMembers(identity.strValue());
+          }
+          Binding updatedBindings = bindingBuilder.build();
+          bindingsList.set(i, updatedBindings);
+          break;
+        }
       }
-      if (identities != null && identities.isEmpty()) {
-        bindings.remove(role);
+
+      Iterator iterator = bindingsList.iterator();
+      while (iterator.hasNext()) {
```

Comment:
If you use `bindingsList.listIterator()` in the loop above, then you could either call `set(updatedBindings)` or `remove()` as needed, and not have this separate loop.